### PR TITLE
Quilt Tooltip API

### PR DESCRIPTION
**Pull request by LambdAurora**
The Quilt Tooltip API is a brand new module in the `gui` library, its main focus is, well... tooltips!

This PR mainly focuses on the `TooltipData`/`TooltipComponent`-related APIs, which to remind everyone this is what the bundle item uses for its fancy tooltip render, other mods have found uses too.

The main issue in Vanilla is `TooltipData` has no "unhardcoded" conversion mechanism to `TooltipComponent`, the only way to convert is by mixining `TooltipComponent#of`, which is rather not practical, historically it was worse with interface mixins being disallowed.

This API introduces 2 new additions:
 - `TooltipComponentCallback`, a straight port of Fabric API's one, use an event-based approach.
 - `ConvertibleTooltipData` which is an extended `TooltipData` interface which adds a **client-only** method to convert it to a `TooltipComponent`. Be careful with it since anyone implementing it **must** use the environment annotation to avoid crashes, this is mentioned heavily in the javadoc. This interface is implemented through the event mentioned earlier, thus it uses the `quilt_tooltip:convertible_tooltip_data` phase which runs before the `quilt:default` phase (which is the default phase for any registered event listener), thus allowing easy overriding through the use of events.

### CODE REVIEW

File: library/gui/tooltip/src/main/java/org/quiltmc/qsl/tooltip/impl/client/QuiltClientTooltipMod.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+	public @Nullable TooltipComponent getComponent(TooltipData data) {
+		if (data instanceof ConvertibleTooltipData convertible) {
+			return convertible.toComponent();
+		}
+
+		return null;
+	}
+
+	static {
+		// We make the convertible tooltip data run before the usual tooltip data conversion listeners to allow overriding through events.
```
Comment (BasiqueEvangelist) [inline library/gui/tooltip/src/main/java/org/quiltmc/qsl/tooltip/impl/client/QuiltClientTooltipMod.java]: If it's run *before*, won't that mean that you can't override it with typical events?

Comment (LambdAurora) [reply library/gui/tooltip/src/main/java/org/quiltmc/qsl/tooltip/impl/client/QuiltClientTooltipMod.java]: Wait...
SHIT

I forgot that it's "who ever returns first wins", damn it.

File: library/gui/tooltip/src/main/java/org/quiltmc/qsl/tooltip/impl/client/QuiltClientTooltipMod.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+		if (data instanceof ConvertibleTooltipData convertible) {
+			return convertible.toComponent();
+		}
+
+		return null;
+	}
+
+	static {
+		// We make the convertible tooltip data runs after the usual tooltip data conversion listeners to allow overriding through events.
+		// (Whoever returns first a tooltip component wins.)
```
Comment (BasiqueEvangelist) [inline library/gui/tooltip/src/main/java/org/quiltmc/qsl/tooltip/impl/client/QuiltClientTooltipMod.java]:
```suggestion
		// (Whoever returns a tooltip component first wins.)
```

File: library/gui/tooltip/src/main/java/org/quiltmc/qsl/tooltip/api/client/TooltipComponentCallback.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+import org.quiltmc.qsl.base.api.event.client.ClientEventAwareListener;
+
+/**
+ * Allows registering a mapping from {@link TooltipData} to {@link TooltipComponent}.
+ * This allows custom tooltips for items: first, override {@link Item#getTooltipData} and return a custom {@link TooltipData}.
+ * Second, register a listener to this event and convert the data to your component implementation if it's an instance of your data class.
+ * <p>
+ * Note that failure to map some data to a component will throw an exception,
+ * so make sure that any data you return in {@link Item#getTooltipData} will be handled by one of the callbacks.
+ */
```
Comment (BasiqueEvangelist) [inline library/gui/tooltip/src/main/java/org/quiltmc/qsl/tooltip/api/client/TooltipComponentCallback.java]: This should also have a reference to the ConvertibleTooltipData helper interface

**Review by LemmaEOF — APPROVED** [review]

File: settings.gradle [hunk 4]
```diff
@@ -28,10 +28,12 @@ includeBuild("build-logic")
 
 // Libraries
 library("core")
+
+library("command")
```
Comment (NoComment1105) [inline settings.gradle]: The rest of this list of libraries *seems* to be following alphabetical order. Not sure if it's coincidence but if it's not, this is out of order

Comment (LambdAurora) [reply settings.gradle]: Damn it me.

**Review by Leo40Git — APPROVED** [review]

File: library/gui/tooltip/src/main/java/org/quiltmc/qsl/tooltip/api/ConvertibleTooltipData.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+ * it will use the {@link org.quiltmc.qsl.tooltip.api.client.TooltipComponentCallback} event with the phase {@link #EVENT_PHASE}.
+ * <p>
+ * <b>WARNING:</b> when implementing this interface,
+ * make sure to annotate the implement {@link #toComponent()} with {@code @Environment(EnvType.CLIENT)}.
+ */
+public interface ConvertibleTooltipData extends TooltipData {
+	/**
+	 * Represents the event phase {@code quilt_tooltip:convertible_tooltip_data}
+	 * which is used when converting this tooltip data to a {@link TooltipComponent},
+	 * it runs after the {@link org.quiltmc.qsl.base.api.event.Event#DEFAULT_PHASE}
```
Comment (TheGlitch76) [inline library/gui/tooltip/src/main/java/org/quiltmc/qsl/tooltip/api/ConvertibleTooltipData.java]: run-on
```suggestion
	 * which is used when converting this tooltip data to a {@link TooltipComponent}.
	 * It runs after the {@link org.quiltmc.qsl.base.api.event.Event#DEFAULT_PHASE}
```

File: library/gui/tooltip/src/main/java/org/quiltmc/qsl/tooltip/api/ConvertibleTooltipData.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+import org.quiltmc.qsl.tooltip.impl.client.QuiltClientTooltipMod;
+
+/**
+ * Represents a {@link TooltipData} which is convertible to a {@link TooltipComponent} on the client.
+ * <p>
+ * When converting a {@link ConvertibleTooltipData} to a {@link TooltipComponent},
+ * it will use the {@link org.quiltmc.qsl.tooltip.api.client.TooltipComponentCallback} event with the phase {@link #EVENT_PHASE}.
+ * <p>
+ * <b>WARNING:</b> when implementing this interface,
+ * make sure to annotate the implement {@link #toComponent()} with {@code @Environment(EnvType.CLIENT)}.
```
Comment (TheGlitch76) [inline library/gui/tooltip/src/main/java/org/quiltmc/qsl/tooltip/api/ConvertibleTooltipData.java]:
```suggestion
 * make sure to annotate the implemented method {@link #toComponent()} with {@code @Environment(EnvType.CLIENT)}.
```

File: library/gui/tooltip/src/main/java/org/quiltmc/qsl/tooltip/api/package-info.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
+/**
+ * <h2>The Quilt Tooltip API</h2>
+ *
+ * <p>
+ * <h3>What are the tooltip APIs?</h3>
+ * With Minecraft 1.17, a new system has been introduced through the {@link net.minecraft.item.BundleItem}: custom tooltip components.
+ * <p>
+ * Those tooltip components are done using mainly two classes:
```
Comment (TheGlitch76) [inline library/gui/tooltip/src/main/java/org/quiltmc/qsl/tooltip/api/package-info.java]: Need a better verb here: created? used? you can't "do" a tooltip :P

File: library/gui/tooltip/src/main/java/org/quiltmc/qsl/tooltip/api/package-info.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+ * <ul>
+ *     <li>{@link net.minecraft.client.item.TooltipData} which will hold information about the tooltip, exist on both sides;</li>
```
Comment (TheGlitch76) [inline library/gui/tooltip/src/main/java/org/quiltmc/qsl/tooltip/api/package-info.java]: if "exist" refers back to the class, it should be "exists", if it refers back to "tooltip", it should be "that exists"

File: library/gui/tooltip/src/main/java/org/quiltmc/qsl/tooltip/api/package-info.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+ *     <li>{@link net.minecraft.client.gui.tooltip.TooltipComponent} which will do the rendering of the tooltip, only exist on the client.</li>
```
Comment (TheGlitch76) [inline library/gui/tooltip/src/main/java/org/quiltmc/qsl/tooltip/api/package-info.java]: Same issue here

File: library/gui/tooltip/src/main/java/org/quiltmc/qsl/tooltip/api/package-info.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+ * </ul>
+ * An item can return a custom tooltip data with its method {@link net.minecraft.item.Item#getTooltipData(net.minecraft.item.ItemStack)}.
+ * The issue is the conversion from the tooltip data to the tooltip component, the tooltip data is not aware of a conversion mechanism.
```
Comment (TheGlitch76) [inline library/gui/tooltip/src/main/java/org/quiltmc/qsl/tooltip/api/package-info.java]: Please check that I understood this correctly.
```suggestion
 * An item can return a custom tooltip data by overriding the method {@link net.minecraft.item.Item#getTooltipData(net.minecraft.item.ItemStack)}.
 * However, there is no method in Vanilla to convert custom tooltip data into a component.
```

**Review by TheGlitch76 — CHANGES_REQUESTED** [review]
API is nice and simple! Just a few comments on javadoc

File: library/gui/tooltip/src/main/java/org/quiltmc/qsl/tooltip/api/package-info.java [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+ * <h2>The Quilt Tooltip API</h2>
+ *
+ * <p>
+ * <h3>What are the tooltip APIs?</h3>
+ * With Minecraft 1.17, a new system has been introduced through the {@link net.minecraft.item.BundleItem}: custom tooltip components.
+ * <p>
+ * Those tooltip components are created using mainly two classes:
+ * <ul>
+ *     <li>{@link net.minecraft.client.item.TooltipData} which will hold information about the tooltip, exists on both sides;</li>
+ *     <li>{@link net.minecraft.client.gui.tooltip.TooltipComponent} which will do the rendering of the tooltip, only exists on the client.</li>
```
Comment (TheGlitch76) [inline library/gui/tooltip/src/main/java/org/quiltmc/qsl/tooltip/api/package-info.java]:
```suggestion
 *     <li>{@link net.minecraft.client.item.TooltipData}, which holds information about the tooltip, exists on both sides;</li>
 *     <li>{@link net.minecraft.client.gui.tooltip.TooltipComponent}, which does the rendering of the tooltip, only exists on the client.</li>
```

**Review by TheGlitch76 — APPROVED** [review]

**Review by BasiqueEvangelist — APPROVED** [review]

**Review by OroArmor — APPROVED** [review]
Maybe add a test for the tooltip callback, otherwise fine

**Comment by LambdAurora** [comment]
> Maybe add a test for the tooltip callback, otherwise fine

The test is already there.

**Comment by OroArmor** [comment]
Oh i see, the new event entrypoint system

**Review by NoComment1105 — APPROVED** [review]
All good 👍🏼

**Pull request merged**